pkg/validations/validate.go

```diff
@@ -12,47 +12,174 @@ import (
+
+func newValidationRunner(threadName string) *validationRunner {
+	return &validationRunner{thread: &starlark.Thread{Name: threadName}, chk: AssertCheck{[]error{}}}
+}
+
+// Visit if `node` has validations in its meta.
+// Runs the validations, any violations from executing the assertions are collected.
+//
+// This visitor stores error(violations) in the validationRunner and returns nil.
+func (a *validationRunner) Visit(node yamlmeta.Node) error {
+	// get rules in node's meta
+	validations := Get(node)
+
+	if validations == nil {
+		return nil
+	}
+	for _, v := range validations {
+		// possible refactor to check validation kwargs prior to validating rules
+		errs := v.Validate(node, a.thread)
+		if errs != nil {
+			a.chk.Violations = append(a.chk.Violations, errs...)
+		}
+	}
+
+	return nil
+}
+
+// Validate runs the assertions in the rules with the node's value as arguments IF
+// the ValidationKwargs conditional options pass.
 //
 // Returns an error if the assertion returns False (not-None), or assert.fail()s.
 // Otherwise, returns nil.
-func (r Rule) Validate(node yamlmeta.Node, thread *starlark.Thread) error {
+func (v NodeValidation) Validate(node yamlmeta.Node, thread *starlark.Thread) []error {
+	key, nodeValue := v.newKeyAndStarlarkValue(node)
+
+	executeRules, err := v.validationKwargs.shouldValidate(nodeValue, thread)
+	if err != nil {
+		return []error{err}
+	}
+	if !executeRules {
+		return nil
+	}
+
+	var failures []error
+	for _, r := range v.rules {
+		result, err := starlark.Call(thread, r.assertion, starlark.Tuple{nodeValue}, []starlark.Tuple{})
+		if err != nil {
+			failures = append(failures, fmt.Errorf("%s (%s) requires %q; %s (by %s)", key, node.GetPosition().AsCompactString(), r.msg, err.Error(), v.position.AsCompactString()))
+		} else if _, ok := result.(starlark.NoneType); !ok {
+			// in order to pass, the assertion must return True or None
+			if !result.Truth() {
+				failures = append(failures, fmt.Errorf("%s (%s) requires %q (by %s)", key, node.GetPosition().AsCompactString(), r.msg, v.position.AsCompactString()))
+			}
+		}
+	}
+	return failures
+}
+
+// shouldValidate uses validationKwargs and the node's value to run checks on the value. If the value satisfies the checks,
+// then the NodeValidation's rules should execute, otherwise the rules will be skipped.
+func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.Thread) (bool, error) {
+	if v.whenNullSkip {
+		if _, ok := value.(starlark.NoneType); ok {
+			return false, nil
+		}
+	}
+
+	if v.when != nil {
+		result, err := starlark.Call(thread, *v.when, starlark.Tuple{value}, []starlark.Tuple{})
+		if err != nil {
+			return false, err
+		}
+
+		switch typedResult := result.(type) {
+		// if None or True, then execute rules
+		case starlark.NoneType:
+			return true, nil
+		case starlark.Bool:
+			return bool(typedResult), nil
+		default:
+			return false, fmt.Errorf("when kwarg function must reutrn a bool or None type")
+		}
```

Comment:
```suggestion
		_, isNone := result.(starlark.NoneType)
		isTrue := bool(result.Truth())
		return isNone || isTrue, nil
```

I'm not seeing a problem in allowing even the value for `when=` to be truth-y (rather than strictly `True`).

Is there?

Reply:
Based on the proposal, there is no difference between the argument to a `when` kwarg and an assertion, so we should handle the return result similarly. 👍